A start at reproducing the benchmarks from the CoNEXT'14 paper

jbapple-cloudera opened the pull request:
The benchmarking code from "Cuckoo Filter: Practically Better Than Bloom", by Fan et al. was apparently lost years ago.

This PR reproduces some of the benchmarks. At this point it includes only benchmarks for the cuckoo filters, not ones for traditional Bloom filters, blocked Bloom filters, Quotient filters, and d-left counting Bloom filters, which were in the paper. The particular implementations of those other filters tested in the CoNEXT'14 paper are not in this repo.

apc999 commented:
@jbapple-cloudera  thanks for the PR! is this PR ready for code review?

jbapple-cloudera commented:
@apc999 Yes, it is ready for code review.

apc999 commented:
Thanks. I will have one pass by Friday. @dave-andersen  you mentioned some on-going work with better hash functions for cuckoo filter, feel free to jump in and provide input. If it is irrelevant to this PR, you can send out your change in separate PR.

dnbh commented:
clhash might also be an appropriate choice as well, and seems to be [faster yet that SIPHash](https://arxiv.org/pdf/1503.03465v8.pdf).

dave-andersen commented:
I favor picking as a default a hash function that can be randomly-keyed to avoid collision problems, and making it easily-templated to accept a faster but less safe hash for situations where the user knows they're not going to be handling potentially-malicious input.  SIPHash covers that, and clhash appears to.

The Dietzfelbinger hash function just pushed doesn't satisfy this, but I think it's a good choice for the short term, because it lets us focus the benchmarking a bit and get something low overhead enough we can easily measure other improvements.  But in the longer term, I'd like to parameterize it.

jbapple-cloudera commented:
I agree with @dave-andersen. In this PR, I was moving to faster hash families partially because they seem to be the main limitation of CF lookup speed. If the benchmark is mainly benchmarking SHA1 or BobHash speed, I didn't think it was effectively showing what the CF can (or can't) do.

dave-andersen commented on benchmarks/conext-table3.cc:
```diff
@@ -0,0 +1,90 @@
+
+  // Insert until failure:
+  size_t inserted = 0;
+  while (inserted < input.size() and 0 == cuckoo.Add(input[inserted])) ++inserted;
```
In keeping with Bin's plan to move everything to the Google C++ style guide, could you switch this (and elsewhere) to use the punctuation operators?  (&& instead of and)

jbapple-cloudera replied:
Done

dave-andersen commented on benchmarks/conext-table3.cc:
```diff
@@ -0,0 +1,90 @@
+const size_t FPR_SAMPLE_SIZE = 1000 * 1000;
+
+struct Metrics {
+  double ndv;    // # of items (million)
```
ndv was unintuitive to me;  value_count?  or something?

jbapple-cloudera replied:
Done

dave-andersen commented on benchmarks/conext-table3.cc:
```diff
@@ -0,0 +1,90 @@
+  // Count false positives
+  size_t false_positive_count = 0;
+  auto i = inserted;
+  for (; i < input.size() and i - inserted < FPR_SAMPLE_SIZE; ++i) {
```
Since i is going to be referenced again 10 lines later, is there a more descriptive name for it?

jbapple-cloudera replied:
Done

dave-andersen commented on benchmarks/conext-table3.cc:
```diff
@@ -0,0 +1,90 @@
+  size_t false_positive_count = 0;
+  auto i = inserted;
+  for (; i < input.size() and i - inserted < FPR_SAMPLE_SIZE; ++i) {
+    false_positive_count += (0 == cuckoo.Contain(input[i]));
```
Note to self:  This awkward construction is a good sign that the API of Cuckoo Filter is broken.  The API should instead be:
  cuckoo.Contains(key) --> bool

This isn't the fault of the current patch, but it's something we should change in a later API overhaul.

dave-andersen commented on benchmarks/random.h:
```diff
@@ -0,0 +1,20 @@
+  ::std::random_device random;
+  auto genrand = [&random]() {
+    return random() + (static_cast<::std::uint64_t>(random()) << 32);
+  };
```
Is there a reason to do it this way instead of:
```
  std::random_device rd;
  std::default_random_engine gen(rd());
  std::uniform_int_distribution<uint64_t> dist;
  std::generate(result.begin(), result.end(), std::bind(dist, gen));
```

jbapple-cloudera replied:
It produces different results.

For the standard library implementation I have installed, default_random is a linear congruential generator; the comments call it "The classic Minimum Standard rand0 of Lewis, Goodman, and Miller." That cuts the false positive rates in half. I believe this is due to my use of the multiply-shift hash function. For a similar example, see Patrascu and Thorup's "The Power of Simple Tabulation Hashing", Figure 2: https://people.csail.mit.edu/mip/papers/charhash/charhash.pdf. It shows that multiply shift behaves better on structured input with some seeds, and much worse with other seeds.

jbapple-cloudera replied:
Done

dave-andersen commented on benchmarks/timing.h:
```diff
@@ -0,0 +1,13 @@
+
+uint64_t Now() {
+  timespec t;
+  if (clock_gettime(CLOCK_MONOTONIC, &t)) {
```
I'm fine with this, but it's worth noting that this won't work on MacOS.  In the longer term, this code should use c++11's high_resolution_clock.  (I'd love it if you'd switch it sooner, of course, because then I could compile on my laptop. :-).  But also fine to get things working and clear this bit up later when we do a portability pass.

jbapple-cloudera replied:
Done

dave-andersen commented on benchmarks/timing.h:
```diff
@@ -0,0 +1,13 @@
+uint64_t Now() {
+  timespec t;
+  if (clock_gettime(CLOCK_MONOTONIC, &t)) {
+    exit(1);
```
`    perror("Could not call clock_gettime"); `

before exit.

jbapple-cloudera replied:
Removed

dave-andersen commented on benchmarks/conext-table3.cc:
```diff
@@ -1,13 +1,13 @@
 // This benchmark reproduces the CoNEXT 2014 results found in "Table 3: Space efficiency
-// and construction speed." It takes about four minutes to run.
+// and construction speed." It takes about two minutes to run.
```
on what CPU type?

dave-andersen commented on benchmarks/conext-figure5.cc:
```diff
@@ -0,0 +1,81 @@
+// This benchmark reproduces the CoNEXT 2014 results found in "Figure 5: Lookup
+// performance when a filter achieves its capacity." It takes about two minutes to run.
+//
+// Results:
```
on what CPU type?

jbapple-cloudera replied:
Done

dave-andersen commented on benchmarks/random.h:
```diff
@@ -0,0 +1,39 @@
+::std::vector<T> MixIn(const T* x_begin, const T* x_end, const T* y_begin, const T* y_end,
+    double y_probability) {
+  const auto x_size = x_end - x_begin, y_size = y_end - y_begin;
+  if (y_size > (1ull << 32)) exit(1);
```
error message on the error exit?

jbapple-cloudera replied:
Done

dave-andersen commented on src/hashutil.h:
```diff
@@ -39,7 +39,11 @@ namespace cuckoofilter {
 
-    private:
+        // See Martin Dietzfelbinger, "Universal hashing and k-wise independent random
+        // variables via integer arithmetic without primes"
```
period at end of comment.

jbapple-cloudera replied:
Done

dave-andersen commented on benchmarks/conext-table3.cc:
```diff
@@ -0,0 +1,90 @@
+
+  auto constr_time = Now() - start_time;
+
+  // Count false positives
```
nit:  Previous and following comments indicated 'this is happening next' with a colon, but this one doesn't match that style.  Keep consistent.

jbapple-cloudera replied:
Done

dave-andersen commented:
Also, because I left it out from my review:  Thanks for doing this!  As you probably noticed, I had just suggested to Bin we needed to do exactly this.  Greatly appreciated!

jbapple-cloudera commented:
When I get a chance to address these, do you want me to rebase/sqaush all the commits in this PR or to add an additional commit with a commit message mentioning that I am addressing code review comments?

dave-andersen commented:
I'm fine with rebasing and squishing unless Bin has a different preference.  But no strong feelings either way - if you want to leave your commit history in there, go for it.

jbapple-cloudera commented:
Rebased and squashed with corrections for code review comments

apc999 commented on benchmarks/conext-table3.cc:
```diff
@@ -0,0 +1,91 @@
+  // Number of distinct values, used only for the constructor of CuckooFilter, which does
+  // not allow the caller to specify the space usage directly. The actual number of
+  // distinct items inserted depends on how many fit until an insert failure occurs.
+  size_t add_count = 127.78 * 1000 * 1000;
```
NIT: `const`?

jbapple-cloudera replied:
Done
